File: chad/src/main/java/org/woahoverflow/chad/commands/gambling/CoinFlip.java
package org.woahoverflow.chad.commands.gambling;

import org.woahoverflow.chad.framework.Util;
import org.woahoverflow.chad.framework.handle.MessageHandler;
import org.woahoverflow.chad.framework.handle.PlayerHandler;
import org.woahoverflow.chad.framework.obj.Command;
import org.woahoverflow.chad.framework.obj.Player;
import org.woahoverflow.chad.framework.obj.Player.DataType;
import sx.blah.discord.handle.impl.events.guild.channel.message.MessageEvent;
import sx.blah.discord.handle.impl.obj.ReactionEmoji;
import sx.blah.discord.handle.obj.IMessage;
import sx.blah.discord.handle.obj.IReaction;
import sx.blah.discord.handle.obj.IUser;
import sx.blah.discord.handle.obj.Permissions;
import sx.blah.discord.util.EmbedBuilder;
import sx.blah.discord.util.RequestBuffer;
import sx.blah.discord.util.RequestBuilder;

import java.security.SecureRandom;
import java.util.HashMap;
import java.util.List;
import java.util.concurrent.TimeUnit;

/**
 * Flips a coin with Chad, or with another user
 *
 * @author sho
 */
public class CoinFlip implements Command.Class{
    @Override
    public final Runnable run(MessageEvent e, List<String> args) {
        return () -> {
            MessageHandler messageHandler = new MessageHandler(e.getChannel(), e.getAuthor());
            Player author = PlayerHandler.handle.getPlayer(e.getAuthor().getLongID());

            if (args.size() == 2 && e.getMessage().getAttachments().isEmpty() && args.get(1).equalsIgnoreCase("tails") || args.get(1).equalsIgnoreCase("heads")) {
                long bet;
                try {
                    bet = Long.parseLong(args.get(0));
                } catch (NumberFormatException throwaway) {
                    messageHandler.sendError("Invalid Bet!");
                    return;
                }

                if (!(bet > 0)) {
                    messageHandler.sendError("Invalid Number!");
                    return;
                }

                long balance = (long) author.getObject(Player.DataType.BALANCE);
                if (bet > balance) {
                    messageHandler.sendError("Your bet is too large!");
                    return;
                }

                if (bet+balance < 0) {
                    messageHandler.sendError("Your balance is too big!\nPlease report this on https://woahoverflow.org/forums");
                    return;
                }

                int user;
                if (args.get(1).equalsIgnoreCase("heads")) {
                    user = 0;
                } else if (args.get(1).equalsIgnoreCase("tails")) {
                    user = 1;
                }
                else {
                    messageHandler.sendError("Please use `heads` or `tails`!");
                    return;
                }

                int flip = new SecureRandom().nextInt(2);

                if (flip == user) {
                    author.setObject(Player.DataType.BALANCE, balance+bet);
                    messageHandler.sendEmbed(new EmbedBuilder().withDesc("You won `"+bet+"`, you now have `" + (balance+bet) + "`."));
                }
                else {
                    author.setObject(Player.DataType.BALANCE, balance-bet);
                    messageHandler.sendEmbed(new EmbedBuilder().withDesc("You lost `"+bet+"`, you now have `" + (balance-bet) + "`."));
                }
            }
            else // assuming that the conditions are met for this
            {
                // Arguments are removed during user building, so I put them here :)
                final String arg0 = args.get(0);

                // Opponent
                IUser opponentUser = null;

                // Builds the user from the arguments
                if (e.getMessage().getMentions().isEmpty()) {
                    StringBuilder stringBuilder = new StringBuilder();
                    args.remove(0);
                    for (String s : args) {
                        stringBuilder.append(s).append(' ');
                        if (!e.getGuild().getUsersByName(stringBuilder.toString().trim()).isEmpty()) {
                            opponentUser = e.getGuild().getUsersByName(stringBuilder.toString().trim()).get(0);
                            break;
                        }
                    }

                    // Checks if the loop actually found a user
                    if (opponentUser == null) {
                        messageHandler.sendError("Invalid User!");
                        return;
                    }
                }
                else {
                    // If there's a mention use that instead
                    if (args.get(1).contains(e.getMessage().getMentions().get(0).getStringID())) {
                        opponentUser = e.getMessage().getMentions().get(0);
                    }
                    else {
                        messageHandler.sendError("Invalid User!");
                        return;
                    }
                }

                // only used once, but thanks lamda
                final IUser user = opponentUser;

                Player opponent = PlayerHandler.handle.getPlayer(opponentUser.getLongID());

                // Sends the invitation message
                IMessage acceptMessage = RequestBuffer.request(() -> e.getChannel().sendMessage("Do you accept `" + e.getAuthor().getName() + "`'s challenge, `" + user.getName() + "`?")).get();

                // Creates a request buffer and reacts with the Y and N emojis
                RequestBuilder rb = new RequestBuilder(e.getClient());
                rb.shouldBufferRequests(true);
                rb.doAction(() -> {
                    acceptMessage.addReaction(ReactionEmoji.of("\uD83C\uDDFE")); // Y
                    return true;
                }).andThen(() -> {
                    acceptMessage.addReaction(ReactionEmoji.of("\uD83C\uDDF3")); // N
                    return true;
                }).execute(); // Executes

                // Assigns variables
                boolean reacted = true;
                int timeout = 0;

                while (reacted) {
                    // If it's been 10 seconds, exit
                    if (timeout == 10) {
                        messageHandler.sendError('`' +user.getName()+"` didn't respond in time!");
                        return;
                    }

                    // Sleeps a second so it doesn't go so fast
                    try {
                        TimeUnit.SECONDS.sleep(1);
                    } catch (InterruptedException e1) {
                        e1.printStackTrace();
                    }

                    // Increases the timeout value
                    timeout++;

                    // Gets both reactions
                    IReaction yReaction = RequestBuffer.request(() -> acceptMessage.getReactionByEmoji(ReactionEmoji.of("\uD83C\uDDFE"))).get();
                    IReaction nReaction = RequestBuffer.request(() -> acceptMessage.getReactionByEmoji(ReactionEmoji.of("\uD83C\uDDF3"))).get();

                    // Checks if the user reacted to the Y
                    if (yReaction.getUserReacted(user)) {
                        reacted = false;
                    }

                    // Checks if the user reacted with the N
                    if (nReaction.getUserReacted(user)) {
                        messageHandler.sendError("User Denied!");
                        return;
                    }
                }

                // Calculates the Bet
                long bet;
                try {
                    bet = Long.parseLong(arg0);
                } catch (NumberFormatException throwaway) {
                    messageHandler.sendError("Invalid Bet!");
                    return;
                }

                if (!(bet > 0)) {
                    messageHandler.sendError("Invalid Number!");
                    return;
                }

                // Gets the author's balance
                long balance = (long) author.getObject(Player.DataType.BALANCE);

                // Checks if the user's bet is bigger than the balance.
                if (bet > balance) {
                    messageHandler.sendError("Your bet is too large!");
                    return;
                }

                // Gets the opponent's balance
                long opponentBalance = (long) opponent.getObject(Player.DataType.BALANCE);

                // Checks if the bet's bigger than the opponent's balance
                if (bet > opponentBalance) {
                    messageHandler.sendError("Your bet is too large for `"+opponentUser.getName()+"`!");
                    return;
                }

                // Checks if the author's balance is too big
                if (bet+balance < 0) {
                    messageHandler.sendError("Your balance is too big!\nPlease report this on https://woahoverflow.org/forums");
                    return;
                }

                // Checks if the opponent's balance is too big
                if (bet+opponentBalance < 0) {
                    messageHandler.sendError('`' +opponentUser.getName()+"`'s balance is too big!\nPlease report this on https://woahoverflow.org/forums");
                    return;
                }

                // Sends a message for the tails & heads declaring
                IMessage pick = RequestBuffer.request(() -> e.getChannel().sendMessage("**X**`heads` or **O**`tails`? __(react)__")).get();

                // Request buffer to apply the reactions
                RequestBuilder r = new RequestBuilder(e.getClient());
                r.shouldBufferRequests(true);

                r.doAction(() -> {
                    pick.addReaction(ReactionEmoji.of("\uD83C\uDDFD")); // Reacts with X
                    return true;
                }).andThen(() -> {
                    pick.addReaction(ReactionEmoji.of("\uD83C\uDDF4")); // Reacts with O
                    return true;
                }).execute(); // Executes the builder

                // Variable declaring
                timeout = 0;
                IUser tails = null;
                IUser heads = null;
                boolean bothReacted = false;
                while (!bothReacted) {
                    // So it doesn't go so fast.
                    try {
                        Thread.sleep(100);
                    } catch (InterruptedException e1) {
                        e1.printStackTrace();
                    }

                    // If the user hasn't responded within 10 seconds, it times out.
                    if (timeout == 10) {
                        messageHandler.sendError('`' +opponentUser.getName()+"` didn't respond in time!");
                        return;
                    }

                    // Adds another second to the timeout
                    timeout++;

                    // X reaction
                    final IReaction x = RequestBuffer.request(() -> pick.getReactionByEmoji(ReactionEmoji.of("\uD83C\uDDFD"))).get();

                    // O reaction
                    final IReaction o = RequestBuffer.request(() -> pick.getReactionByEmoji(ReactionEmoji.of("\uD83C\uDDF4"))).get();

                    if (heads == null) {
                        if (o.getUserReacted(e.getAuthor())) {
                            if (tails == null)
                                heads = e.getAuthor();
                            else if (!tails.equals(e.getAuthor()))
                                heads = e.getAuthor();
                        }

                        if (o.getUserReacted(opponentUser)) {
                            if (tails == null)
                                heads = opponentUser;
                            else if (!tails.equals(opponentUser))
                                heads = opponentUser;
                        }
                    }

                    if (tails == null) {
                        if (x.getUserReacted(e.getAuthor())) {
                            if (heads == null)
                                tails = e.getAuthor();
                            else if (!heads.equals(e.getAuthor()))
                                tails = e.getAuthor();
                        }

                        if (x.getUserReacted(opponentUser)) {
                            if (heads == null)
                                tails = opponentUser;
                            else if (!heads.equals(opponentUser))
                                tails = opponentUser;
                        }
                    }


                    // If both users have selected one, the loop stops.
                    if (tails != null & heads != null)
                        bothReacted = true;
                }

                // Removes all the reactions
                if (e.getClient().getOurUser().getPermissionsForGuild(e.getGuild()).contains(Permissions.MANAGE_MESSAGES))
                    RequestBuffer.request(pick::removeAllReactions);

                // Ties the user's balances to their name
                long tailsBalance;
                long headsBalance;
                if (e.getAuthor().equals(tails) && user.equals(heads)) {
                    tailsBalance = balance;
                    headsBalance = opponentBalance;
                }
                else {
                    tailsBalance = opponentBalance;
                    headsBalance = balance;
                }

                // 0 is tails winning, 1 is heads winning
                if (Util.coinFlip()) {
                    // Sets the user's balances
                    PlayerHandler.handle.getPlayer(tails.getLongID()).setObject(DataType.BALANCE, tailsBalance+bet);
                    PlayerHandler.handle.getPlayer(heads.getLongID()).setObject(DataType.BALANCE, tailsBalance-bet);

                    // Creates the edit string, then applies.
                    final String editString = '`' +tails.getName()+"` has won `" + bet + "`!"
                        + "\n\n`"+tails.getName()+"` now has `"+(tailsBalance+bet)+"`, `"+heads.getName()+"` now has `"+(headsBalance-bet)+ '`';
                    RequestBuffer.request(() -> pick.edit(editString));
                }
                else /* flip is 1, so heads wins this */ {
                    // Sets the user's balances
                    PlayerHandler.handle.getPlayer(tails.getLongID()).setObject(DataType.BALANCE, tailsBalance-bet);
                    PlayerHandler.handle.getPlayer(heads.getLongID()).setObject(DataType.BALANCE, tailsBalance+bet);

                    // Creates the edit string, then applies.
                    final String editString = '`' +heads.getName()+"` has won `" + bet + "`!"
                        + "\n`"+heads.getName()+"` now has `"+(headsBalance+bet)+"`, `"+tails.getName()+"` now has `"+(tailsBalance-bet)+ '`';
                    RequestBuffer.request(() -> pick.edit(editString));
                }
            }
        };
    }

    @Override
    public final Runnable help(MessageEvent e) {
        HashMap<String, String> st = new HashMap<>();
        st.put("coinflip <amount to bet> <tails/heads>", "Play CoinFlip with Chad");
        st.put("coinflip <amount to bet> <@user>", "Play CoinFlip with another user");
        return Command.helpCommand(st, "CoinFlip", e);
    }
}

File: chad/src/main/java/org/woahoverflow/chad/commands/gambling/CoinFlip.kt
package org.woahoverflow.chad.commands.gambling

import kotlinx.coroutines.delay
import org.woahoverflow.chad.framework.handle.GuildHandler
import org.woahoverflow.chad.framework.handle.MessageHandler
import org.woahoverflow.chad.framework.handle.PlayerHandler
import org.woahoverflow.chad.framework.handle.coroutine.isUnit
import org.woahoverflow.chad.framework.handle.coroutine.request
import org.woahoverflow.chad.framework.obj.Command
import org.woahoverflow.chad.framework.obj.Guild
import org.woahoverflow.chad.framework.obj.Player.DataType
import org.woahoverflow.chad.framework.util.validBet
import sx.blah.discord.handle.impl.events.guild.channel.message.MessageEvent
import sx.blah.discord.handle.impl.obj.ReactionEmoji
import sx.blah.discord.handle.obj.IMessage
import sx.blah.discord.handle.obj.IReaction
import sx.blah.discord.handle.obj.IUser
import sx.blah.discord.util.EmbedBuilder
import sx.blah.discord.util.RequestBuffer
import sx.blah.discord.util.RequestBuilder
import java.util.*
import kotlin.random.Random

/**
 * Flips a coin with Chad, or with another user
 *
 * @author sho
 */
class CoinFlip : Command.Class {
    override suspend fun run(e: MessageEvent, args: MutableList<String>) {
        val messageHandler = MessageHandler(e.channel, e.author)
        val author = PlayerHandler.getPlayer(e.author.longID)
        val prefix = GuildHandler.getGuild(e.guild.longID).getObject(Guild.DataType.PREFIX) as String

        if (e.message.mentions.isEmpty() && args.size == 2) {
            // Makes sure the bot is a valid long
            val bet: Long = try {
                args[0].toLong()
            } catch (ex: NumberFormatException) {
                messageHandler.sendError("Invalid Bet!")
                return
            }

            val balance = author.getObject(DataType.BALANCE) as Long
            if (!validBet(bet, balance)) {
                messageHandler.sendError("Invalid Bet!")
                return
            }

            val user: Int = when {
                args[1].equals("heads", ignoreCase = true) -> 0
                args[1].equals("tails", ignoreCase = true) -> 1

                else -> {
                    messageHandler.sendError("Please use `heads` or `tails`!")
                    return
                }
            }

            val flip = Random.nextInt(2)

            if (flip == user) {
                author.setObject(DataType.BALANCE, balance + bet)
                messageHandler.sendEmbed(EmbedBuilder().withDesc("You won `$bet`, you now have `${balance+bet}`!"))
            } else {
                author.setObject(DataType.BALANCE, balance - bet)
                messageHandler.sendEmbed(EmbedBuilder().withDesc("You lost `$bet`, you now have `${balance-bet}`!"))
            }
            return
        } else if (e.message.mentions.size == 1 && args.size == 2) {
            val user: IUser = if (e.message.mentions.isNotEmpty() && args[1].contains(e.message.mentions[0].stringID)) {
                e.message.mentions[0]
            } else {
                messageHandler.sendPresetError(MessageHandler.Messages.INVALID_ARGUMENTS, "${prefix}coinflip [@user] [bet]")
                return
            }

            if (user == e.author || user == e.client.ourUser) {
                messageHandler.sendPresetError(MessageHandler.Messages.INVALID_ARGUMENTS, "${prefix}coinflip [@user] [bet]")
                return
            }

            val bet: Long = try {
                args[0].toLong()
            } catch (ex: NumberFormatException) {
                messageHandler.sendError("Invalid Bet!")
                return
            }

            val acceptMessage = request {
                e.channel.sendMessage("Do you accept `${e.author.name}`'s challenge, `${user.name}`?")
            }.also {
                if (it.isUnit() || it.result !is IMessage)
                    return messageHandler.sendPresetError(MessageHandler.Messages.INTERNAL_EXCEPTION)
            }.result as IMessage

            val rb = RequestBuilder(e.client)
            rb.shouldBufferRequests(true)
            rb.doAction {
                acceptMessage.addReaction(ReactionEmoji.of("\uD83C\uDDFE")) // Y
                true
            }.andThen {
                acceptMessage.addReaction(ReactionEmoji.of("\uD83C\uDDF3")) // N
                true
            }.execute() // Executes

            var reacted = false
            var timeout = 0

            while (!reacted) {
                if (timeout == 10) {
                    messageHandler.sendError("`${user.name}` didn't respond in time!")
                    return
                }

                delay(1000L)

                timeout++

                val yReaction = request {
                    acceptMessage.getReactionByEmoji(ReactionEmoji.of("\uD83C\uDDFE"))
                }.also {
                    if (it.isUnit() || it.result !is IReaction)
                        return messageHandler.sendPresetError(MessageHandler.Messages.INTERNAL_EXCEPTION)
                }.result as IReaction

                val nReaction = request {
                    acceptMessage.getReactionByEmoji(ReactionEmoji.of("\uD83C\uDDF3"))
                }.also {
                    if (it.isUnit() || it.result !is IReaction)
                        return messageHandler.sendPresetError(MessageHandler.Messages.INTERNAL_EXCEPTION)
                }.result as IReaction

                if (yReaction.getUserReacted(user)) reacted = true

                if (nReaction.getUserReacted(user)) {
                    messageHandler.sendError("`${user.name}` denied!")
                    return
                }
            }

            val otherUser = PlayerHandler.getPlayer(user.longID)

            val otherBalance = otherUser.getObject(DataType.BALANCE) as Long
            val balance = author.getObject(DataType.BALANCE) as Long
            if (!validBet(bet, balance, otherBalance = otherBalance)) {
                messageHandler.sendError("Invalid Bet!")
                return
            }

            val pick = request {
                e.channel.sendMessage("**X** TAILS\n**O** HEADS")
            }.also {
                if (it.isUnit() || it.result !is IMessage)
                    return messageHandler.sendPresetError(MessageHandler.Messages.INTERNAL_EXCEPTION)
            }.result as IMessage

            // Request buffer to apply the reactions
            val r = RequestBuilder(e.client)
            r.shouldBufferRequests(true)

            r.doAction {
                pick.addReaction(ReactionEmoji.of("\uD83C\uDDFD")) // X
                true
            }.andThen {
                pick.addReaction(ReactionEmoji.of("\uD83C\uDDF4")) // O
                true
            }.execute()

            timeout = 0
            var tails: IUser? = null
            var heads: IUser? = null
            while (tails == null || heads == null) {
                delay(10000L)

                if (timeout * 100 * 10 > 10) {
                    messageHandler.sendError("`${user.name}` or `${e.author.name}` didn't respond in time!")
                    return
                }

                // X reaction
                val x = request {
                    pick.getReactionByEmoji(ReactionEmoji.of("\uD83C\uDDFD"))
                }.also {
                    if (it.isUnit() || it.result !is IReaction)
                        return messageHandler.sendPresetError(MessageHandler.Messages.INTERNAL_EXCEPTION)
                }.result as IReaction

                // O reaction
                val o = request {
                    pick.getReactionByEmoji(ReactionEmoji.of("\uD83C\uDDF4"))
                }.also {
                    if (it.isUnit() || it.result !is IReaction)
                        return messageHandler.sendPresetError(MessageHandler.Messages.INTERNAL_EXCEPTION)
                }.result as IReaction

                if (tails == null) {
                    if (x.getUserReacted(user)) {
                        if (heads != user) tails = user
                    }

                    if (x.getUserReacted(e.author)) {
                        if (heads != e.author) tails = e.author
                    }
                }

                if (heads == null) {
                    if (o.getUserReacted(user)) {
                        if (tails != user) heads = user
                    }

                    if (o.getUserReacted(e.author)) {
                        if (tails != e.author) heads = e.author
                    }
                }
            }

            RequestBuffer.request {
                pick.delete()
            }

            val tailsBalance: Long = if (e.author == tails) balance else otherBalance
            val headsBalance: Long = if (e.author == heads) balance else otherBalance

            val flip = Random.nextInt(2)

            // 0 = Tails; 1 = Heads
            if (flip == 0) {
                PlayerHandler.getPlayer(tails.longID).setObject(DataType.BALANCE, tailsBalance + bet)
                PlayerHandler.getPlayer(heads.longID).setObject(DataType.BALANCE, headsBalance - bet)

                messageHandler.sendMessage("`${tails.name}` has won `$bet`!\n\n" +
                        "`${tails.name}` now has `${tailsBalance + bet}`, and `${heads.name}` now has `${headsBalance - bet}`.")
            } else {
                PlayerHandler.getPlayer(tails.longID).setObject(DataType.BALANCE, tailsBalance - bet)
                PlayerHandler.getPlayer(heads.longID).setObject(DataType.BALANCE, headsBalance + bet)

                messageHandler.sendMessage("`${heads.name}` has won `$bet`!\n\n" +
                        "`${heads.name}` now has `${headsBalance + bet}`, and `${tails.name}` now has `${tailsBalance - bet}`.")

            }
        } else {
            messageHandler.sendPresetError(MessageHandler.Messages.INVALID_ARGUMENTS, "${prefix}coinflip [amount to bet] [@user] / ${prefix}coinflip [amount to bet] [tails/head]")
        }
    }

    override suspend fun help(e: MessageEvent) {
        val st = HashMap<String, String>()
        st["coinflip [amount to bet] [tails/heads]"] = "Play CoinFlip with Chad"
        st["coinflip [amount to bet] [@user]"] = "Play CoinFlip with another user"
        Command.helpCommand(st, "CoinFlip", e)
    }
}
